src/main/java/com/zendesk/maxwell/producer/MaxwellRedisProducer.java

```diff
@@ -6,42 +6,98 @@
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 import redis.clients.jedis.Jedis;
+import redis.clients.jedis.StreamEntryID;
 import redis.clients.jedis.exceptions.JedisConnectionException;
+import java.util.HashMap;
+import java.util.Map;
 
 public class MaxwellRedisProducer extends AbstractProducer implements StoppableTask {
 	private static final Logger logger = LoggerFactory.getLogger(MaxwellRedisProducer.class);
 	private final String channel;
-	private final String listkey;
-	private final String redistype;
+	private final String redisType;
 	private final Jedis jedis;
 
-	public MaxwellRedisProducer(MaxwellContext context, String redisPubChannel, String redisListKey, String redisType) {
+	@Deprecated
+    public MaxwellRedisProducer(MaxwellContext context, String redisPubChannel, String redisListKey, String redisType) {
+		this(context, redisType);
+	}
+
+	public MaxwellRedisProducer(MaxwellContext context, String redisType) {
 		super(context);
 
-		channel = redisPubChannel;
-		listkey = redisListKey;
-		redistype = redisType;
+		if (this.context.getConfig().redisListKey != null) {
+			channel = context.getConfig().redisListKey;
+		}
+		else if (this.context.getConfig().redisStreamKey != null) {
+			channel = context.getConfig().redisStreamKey;
+		}
+		else {
+			channel = this.context.getConfig().redisPubChannel;
+		}
+
+		this.redisType = redisType;
 
 		jedis = new Jedis(context.getConfig().redisHost, context.getConfig().redisPort);
 		jedis.connect();
+
 		if (context.getConfig().redisAuth != null) {
 			jedis.auth(context.getConfig().redisAuth);
 		}
+
 		if (context.getConfig().redisDatabase > 0) {
 			jedis.select(context.getConfig().redisDatabase);
 		}
 	}
 
-	private void sendToRedis(String msg) {
-		switch (redistype) {
+	private void sendToRedis(RowMap msg) throws Exception {
+		String messageStr = msg.toJSON(outputConfig);
+
+		switch (redisType) {
 			case "lpush":
-				jedis.lpush(this.listkey, msg);
+				jedis.lpush(this.channel, messageStr);
+				break;
+			case "xadd":
+				Map<String, String> message = new HashMap<>();
+
+				String jsonKey = this.context.getConfig().redisStreamJsonKey;
+				
+				if (jsonKey == null) {
+					// TODO dot notated map impl in RowMap.toJson
+					throw new IllegalArgumentException("Stream requires key name for serialized JSON value");
+				}
+				else {
+					message.put(jsonKey, messageStr);
```

Comment:
just curious; is there a reason we're passing the whole thing as a JSON blob rather than using the internal redis KV store that's in the message?

Reply:
Since you're curious-

I have a use case where the JSON body is the value part of the first K/V pair and nothing more. So I'll justify why support JSON body as first K/V pair instead of flattened K/V pair from a Maxwell integrator POV.

In practice, an integration with a Stream or Queue will introduce an interface that eliminates the concern of message source to message consumers. What if I wrote code ready for the case: No Redis. Yes Kafka?

For example, porting an implementation of Maxwell messages from list (used as queue) to stream is almost as simple as replacing a few Redis commands since a JSON body is to be read either way.

Flatting creates the same result. In production already, it should be a backwards-compatible feature as K/V (dot-notated?) flat mapping will [preferably] obsolete this practice. It requires implementing expansion and contraction on both the message's touch points: Maxwell and the Consumer via dot-notated keys, which is not (as of writing) implemented in client libraries anyway. This implementation path is most likely, and should be supported first to obtain the benefits of Redis streams.

Myself, I have yet to check the performance differences between flattened K/V and "json(value[0])" - so I know the performance of the decoding of JSON but not the impact on performance of the underlying data storage both redis, and protocol-compatible, disk-based K/V store like WiredTiger or RocksDB (implemented through ardb).


All that being said, this can be supported with little maintenance.